R/teal_lockfile.R

```diff
@@ -23,92 +44,139 @@
 #'
 #' @seealso [renv::snapshot()], [renv::restore()].
 #'
-#' @return Nothing. This function is executed for its side effect of creating a lockfile used in the `teal` application.
+#' @return
+#'  `ExtendedTask` processing `renv` lockfile or `NULL` if lockfile has been provided through options
+#'  (skipping asynchronous process).
+#'
+#' @name teal_lockfile
+#' @rdname teal_lockfile
 #'
 #' @keywords internal
-teal_lockfile <- function() {
+NULL
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_invoke <- function() {
   lockfile_path <- "teal_app.lock"
+  shiny::onStop(function() file.remove(lockfile_path))
+
   # If user has setup the file, there is no need to compute a new one.
   user_lockfile <- getOption("teal.renv.lockfile", "")
   if (!identical(user_lockfile, "")) {
     if (file.exists(user_lockfile)) {
       file.copy(user_lockfile, lockfile_path)
-      return(invisible(NULL))
+      logger::log_trace('Lockfile set using option "teal.renv.lockfile" - skipping automatic creation.')
+      return(NULL)
     } else {
       stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
     }
   }
 
-  if (!(is_in_test() || is_r_cmd_check())) {
-    old_plan <- future::plan()
-    # If there is already a parallel (non-sequential) backend, reuse it.
-    if (inherits(old_plan, "sequential")) {
-      future::plan(future::multisession, workers = 2)
+  process <- ExtendedTask$new(
+    function(run, lockfile_path = lockfile_path, opts = opts, sysenv = sysenv, libpaths = libpaths, wd = wd) {
+      mirai::mirai(
+        run(lockfile_path = lockfile_path, opts = opts, sysenv = sysenv, libpaths = libpaths, wd = wd),
+        run = run, lockfile_path = lockfile_path, opts = opts, sysenv = sysenv, libpaths = libpaths, wd = wd
+      )
     }
+  )
 
-    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
-    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
-    logger::log_trace("lockfile creation invoked.")
-  }
+  suppressWarnings({ # 'package:stats' may not be available when loading
+    process$invoke(
+      run = teal:::renv_snapshot,
```

Comment:
Remember, CRAN doesn't take kindly to using `:::`.
It's unclear to my why it is neede, though, so maybe you have exhausted other options.

Reply:
Yeah, there is even our own R CMD CHECK Notes blocker that does not allow that

```r
* checking dependencies in R code ... NOTE
There are ::: calls to the package's namespace in its code. A package
  almost never needs to use ::: for its own objects:
  ‘renv_snapshot’
```

I was trying to overcome different issue with https://github.com/insightsengineering/teal/pull/1263#issuecomment-2230583352
Let's try if we can pass this directly without the specs of the namespace